merlin/table/tensor_column.py

```diff
@@ -0,0 +1,155 @@
+    """
+
+    @classmethod
+    @abstractmethod
+    def array_type(cls) -> Type:
+        """
+        The type of the arrays backing this column
+        """
+        return None
+
+    @classmethod
+    @abstractmethod
+    def supported_devices(cls):
+        """
+        List of device types supported by this column type
+        """
+        return []
+
+    def __init__(self, values: Any, offsets: Any = None, dtype=None, _ref=None, _device=None):
+        self._validate_values_offsets(values, offsets)
+
+        self._values = values
+        self._offsets = offsets
+
+        self._dtype = md.dtype(dtype or values.dtype)
+        self._ref = _ref
+        self._device = _device
+
+    @property
+    def device(self) -> Device:
+        return self._device
+
+    @property
+    def values(self):
+        return self._values
+
+    @property
+    def offsets(self):
+        return self._offsets
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    def __getitem__(self, index):
+        # get correct indexes if offsets exists
+        if self._offsets is None:
+            index = len(self._values) + index if index < 0 else index
+            return self._values[index]
+        # There should be a better way to get negative indexing to work
+        index = len(self._offsets) - 1 + index if index < 0 else index
+        start = self._offsets[index]
+        end = self._offsets[index + 1]
+        return self._values[start:end]
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return False
+        return (
+            _arrays_eq(self._values, other._values)
+            and _arrays_eq(self._offsets, other._offsets)
+            and self.dtype == other.dtype
+            and self.device == other.device
+        )
+
+    def _validate_values_offsets(self, values, offsets):
+        self._raise_type_error("values", values)
+
+        if offsets is not None:
+            self._raise_type_error("offsets", offsets)
+
+            # The offsets have to be in increasing order
+            # None of the offsets should be greater than the length of the values
+            # The last offset should be the length of the values
+
+            if len(values) != offsets[-1]:
+                raise ValueError("The last offset must match the length of the values.")
+
+            if any(offsets[1:] - offsets[:-1] < 0):
+                raise ValueError("Offsets must be in increasing order")
+
+    def _raise_type_error(self, field_name, array):
+        expected_type = self.__class__.array_type()
+        if not isinstance(  # pylint:disable=isinstance-second-argument-not-valid-type
+            array, expected_type
+        ):
+            raise TypeError(
+                f"{self.__class__} expected arrays of type {expected_type}, "
+                f"but received {field_name} of type {type(array)}"
+            )
+
+
+@dataclass(frozen=True)
+class TransferColumn:
```

Comment:
Wonder if this class and the sub-classes below should be semi-private and have names starting with leading underscores?

Reply:
Ok I went ahead and made this change. I also removed all of the array based logic, since we are using dlpack only.